File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/appwidget/ListWidgetPreferenceFragment.java
package com.battlelancer.seriesguide.appwidget;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceCategory;
import android.preference.PreferenceFragment;
import android.preference.PreferenceGroup;
import android.preference.PreferenceScreen;
import android.support.annotation.ArrayRes;
import android.support.annotation.StringRes;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.settings.WidgetSettings;
import com.battlelancer.seriesguide.util.Utils;

/**
 * Shows settings fragment for a specific app widget, hosted inside a {@link ListWidgetConfigure}
 * activity.
 */
public class ListWidgetPreferenceFragment extends PreferenceFragment {

    private PreferenceScreen preferenceScreen;
    private ListPreference typePref;
    private ListPreference showsSortPref;
    private CheckBoxPreference onlyCollectedPref;
    private CheckBoxPreference hideWatchedPreference;
    private CheckBoxPreference isInfinitePref;

    public static ListWidgetPreferenceFragment newInstance(int appWidgetId) {
        ListWidgetPreferenceFragment f = new ListWidgetPreferenceFragment();

        Bundle args = new Bundle();
        args.putInt("appWidgetId", appWidgetId);
        f.setArguments(args);

        return f;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        int appWidgetId = getArguments().getInt("appWidgetId");

        typePref = listPref(
                WidgetSettings.KEY_PREFIX_WIDGET_LISTTYPE + appWidgetId,
                R.string.pref_widget_type,
                R.array.widgetType,
                R.array.widgetTypeData,
                getString(R.string.widget_default_type)
        );
        ListPreference themePref = listPref(
                WidgetSettings.KEY_PREFIX_WIDGET_THEME + appWidgetId,
                R.string.pref_theme,
                R.array.widgetTheme,
                R.array.widgetThemeData,
                "0"
        );
        ListPreference backgroundPref = listPref(
                WidgetSettings.KEY_PREFIX_WIDGET_BACKGROUND_OPACITY + appWidgetId,
                R.string.pref_widget_opacity,
                R.array.widgetOpacity,
                R.array.widgetOpacityData,
                WidgetSettings.DEFAULT_WIDGET_BACKGROUND_OPACITY
        );

        showsSortPref = listPref(
                WidgetSettings.KEY_PREFIX_WIDGET_SHOWS_SORT_ORDER + appWidgetId,
                R.string.action_shows_sort,
                R.array.widgetShowSortOrder,
                R.array.widgetShowSortOrderData,
                getString(R.string.widget_default_show_sort_order)
        );

        CheckBoxPreference onlyFavoritesPref = checkBoxPref(
                WidgetSettings.KEY_PREFIX_WIDGET_ONLY_FAVORITES + appWidgetId,
                R.string.only_favorites,
                false
        );
        onlyCollectedPref = checkBoxPref(
                WidgetSettings.KEY_PREFIX_WIDGET_ONLY_COLLECTED + appWidgetId,
                R.string.calendar_only_collected,
                false
        );
        hideWatchedPreference = checkBoxPref(
                WidgetSettings.KEY_PREFIX_WIDGET_HIDE_WATCHED + appWidgetId,
                R.string.hide_watched,
                true
        );
        isInfinitePref = checkBoxPref(
                WidgetSettings.KEY_PREFIX_WIDGET_IS_INFINITE + appWidgetId,
                R.string.pref_infinite_scrolling,
                false
        );

        // use the settings file specific to widgets
        getPreferenceManager().setSharedPreferencesName(WidgetSettings.SETTINGS_FILE);
        getPreferenceManager().setSharedPreferencesMode(0);

        // create a widget specific settings screen
        preferenceScreen = getPreferenceManager().createPreferenceScreen(getActivity());

        preferenceScreen.addPreference(typePref);
        preferenceScreen.addPreference(showsSortPref);
        preferenceScreen.addPreference(onlyFavoritesPref);
        preferenceScreen.addPreference(onlyCollectedPref);
        preferenceScreen.addPreference(hideWatchedPreference);
        preferenceScreen.addPreference(isInfinitePref);

        PreferenceCategory appearanceCategory = new PreferenceCategory(getActivity());
        appearanceCategory.setTitle(R.string.pref_appearance);
        preferenceScreen.addPreference(appearanceCategory);
        appearanceCategory.addPreference(themePref);
        appearanceCategory.addPreference(backgroundPref);

        setPreferenceScreen(preferenceScreen);

        // after restoring any values prevent persisting changes (only save on user command)
        disablePersistence(preferenceScreen);

        bindPreferenceSummaryToValue(typePref);
        bindPreferenceSummaryToValue(showsSortPref);
        bindPreferenceSummaryToValue(backgroundPref);
        bindPreferenceSummaryToValue(themePref);

        if (!Utils.hasAccessToX(getActivity())) {
            // disable saving prefs not available for non-supporters
            Preference.OnPreferenceChangeListener onDisablePreferenceChangeListener
                    = new Preference.OnPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object newValue) {
                    Utils.advertiseSubscription(getActivity());
                    return false;
                }
            };
            typePref.setOnPreferenceChangeListener(onDisablePreferenceChangeListener);
            typePref.setSummary(R.string.onlyx);
            themePref.setOnPreferenceChangeListener(onDisablePreferenceChangeListener);
            themePref.setSummary(R.string.onlyx);
            backgroundPref.setOnPreferenceChangeListener(onDisablePreferenceChangeListener);
            backgroundPref.setSummary(R.string.onlyx);
        }
    }

    @Override
    public void onResume() {
        super.onResume();

        // trigger the listener to handle the current state
        preferenceChangeListener.onPreferenceChange(typePref, typePref.getValue());
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.widget_config_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.menu_save) {
            save(); // disabled persistence, so save ourselves
            ((ListWidgetConfigure) getActivity()).updateWidget();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private CheckBoxPreference checkBoxPref(String key, @StringRes int titleRes,
            boolean defaultValue) {
        CheckBoxPreference checkBoxPreference = new CheckBoxPreference(getActivity());
        checkBoxPreference.setKey(key);
        checkBoxPreference.setTitle(titleRes);
        checkBoxPreference.setDefaultValue(defaultValue);
        return checkBoxPreference;
    }

    private ListPreference listPref(String key, @StringRes int title, @ArrayRes int entries,
            @ArrayRes int values, String defaultValue) {
        ListPreference listPreference = new ListPreference(getActivity());
        listPreference.setKey(key);
        listPreference.setTitle(title);
        listPreference.setEntries(entries);
        listPreference.setEntryValues(values);
        listPreference.setDefaultValue(defaultValue);
        listPreference.setPositiveButtonText(null);
        listPreference.setNegativeButtonText(null);
        return listPreference;
    }

    private void disablePersistence(PreferenceGroup group) {
        int count = group.getPreferenceCount();
        for (int i = 0; i < count; i++) {
            Preference pref = group.getPreference(i);
            if (pref instanceof PreferenceGroup) {
                disablePersistence((PreferenceGroup) pref);
            } else {
                pref.setPersistent(false);
            }
        }
    }

    private void save() {
        SharedPreferences.Editor editor = getPreferenceManager().getSharedPreferences().edit();
        savePreferences(preferenceScreen, editor);
        editor.apply();
    }

    private void savePreferences(PreferenceGroup group, SharedPreferences.Editor editor) {
        int count = group.getPreferenceCount();
        for (int i = 0; i < count; i++) {
            Preference pref = group.getPreference(i);
            String key = pref.getKey();
            if (pref instanceof PreferenceGroup) {
                savePreferences((PreferenceGroup) pref, editor);
            } else if (pref instanceof CheckBoxPreference) {
                editor.putBoolean(key, ((CheckBoxPreference) pref).isChecked());
            } else if (pref instanceof ListPreference) {
                editor.putString(key, ((ListPreference) pref).getValue());
            }
        }
    }

    /**
     * Binds a preference's summary to its value. More specifically, when the
     * preference's value is changed, its summary (line of text below the
     * preference title) is updated to reflect the value. The summary is also
     * immediately updated upon calling this method. The exact display format is
     * dependent on the type of preference.
     */
    public void bindPreferenceSummaryToValue(ListPreference preference) {
        // Set the listener to watch for value changes.
        preference.setOnPreferenceChangeListener(preferenceChangeListener);

        // Trigger the listener immediately with the preference's current value.
        preferenceChangeListener.onPreferenceChange(preference, preference.getValue());
    }

    private Preference.OnPreferenceChangeListener preferenceChangeListener =
            new Preference.OnPreferenceChangeListener() {
                @Override
                public boolean onPreferenceChange(Preference preference, Object newValue) {
                    // disable episode related setting if selecting show widget type
                    if (typePref.getKey().equals(preference.getKey())) {
                        String newTypeValue = (String) newValue;
                        boolean displayingShows = getString(R.string.widget_type_shows)
                                .equals(newTypeValue);
                        showsSortPref.setEnabled(displayingShows);
                        onlyCollectedPref.setEnabled(!displayingShows);
                        hideWatchedPreference.setEnabled(!displayingShows);
                        isInfinitePref.setEnabled(!displayingShows);
                    }

                    if (preference instanceof ListPreference) {
                        // For list preferences, look up the correct display value in
                        // the preference's 'entries' list.
                        ListPreference listPreference = (ListPreference) preference;
                        String stringValue = newValue.toString();
                        int index = listPreference.findIndexOfValue(stringValue);

                        // Set the summary to reflect the new value. Escape '%'.
                        preference.setSummary(
                                index >= 0
                                        ? (listPreference.getEntries()[index])
                                        .toString().replaceAll("%", "%%")
                                        : null
                        );
                    }

                    return true;
                }
            };
}

File: app/src/main/java/com/battlelancer/seriesguide/appwidget/ListWidgetPreferenceFragment.kt
package com.battlelancer.seriesguide.appwidget

import android.content.SharedPreferences
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import androidx.annotation.ArrayRes
import androidx.annotation.StringRes
import androidx.core.content.edit
import androidx.core.os.bundleOf
import androidx.preference.CheckBoxPreference
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceCategory
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceGroup
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.settings.WidgetSettings
import com.battlelancer.seriesguide.util.Utils

/**
 * Shows settings fragment for a specific app widget, hosted inside a [ListWidgetPreferenceActivity]
 * activity.
 */
class ListWidgetPreferenceFragment : PreferenceFragmentCompat() {

    private lateinit var typePref: ListPreference
    private lateinit var showsSortPref: ListPreference
    private lateinit var onlyPremieresPref: CheckBoxPreference
    private lateinit var onlyCollectedPref: CheckBoxPreference
    private lateinit var hideWatchedPreference: CheckBoxPreference
    private lateinit var isInfinitePref: CheckBoxPreference
    private lateinit var themePref: ListPreference
    private lateinit var backgroundPref: ListPreference

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreatePreferences(
        savedInstanceState: Bundle?,
        rootKey: String?
    ) {
        val appWidgetId = requireArguments().getInt(ARG_WIDGET_ID)

        // Type.
        typePref = createListPref(
            WidgetSettings.KEY_PREFIX_WIDGET_LISTTYPE + appWidgetId,
            R.string.pref_widget_type,
            R.array.widgetType,
            R.array.widgetTypeData,
            getString(R.string.widget_default_type)
        )

        // Shows only sort order.
        showsSortPref = createListPref(
            WidgetSettings.KEY_PREFIX_WIDGET_SHOWS_SORT_ORDER + appWidgetId,
            R.string.action_shows_sort,
            R.array.widgetShowSortOrder,
            R.array.widgetShowSortOrderData,
            getString(R.string.widget_default_show_sort_order)
        )

        // Show/episode favorites filter.
        val onlyFavoritesPref = createCheckBoxPref(
            WidgetSettings.KEY_PREFIX_WIDGET_ONLY_FAVORITES + appWidgetId,
            R.string.only_favorites,
            false
        )

        // Episode only filters.
        onlyPremieresPref = createCheckBoxPref(
            WidgetSettings.KEY_PREFIX_WIDGET_ONLY_PREMIERES + appWidgetId,
            R.string.calendar_only_premieres,
            false
        )
        onlyCollectedPref = createCheckBoxPref(
            WidgetSettings.KEY_PREFIX_WIDGET_ONLY_COLLECTED + appWidgetId,
            R.string.calendar_only_collected,
            false
        )
        hideWatchedPreference = createCheckBoxPref(
            WidgetSettings.KEY_PREFIX_WIDGET_HIDE_WATCHED + appWidgetId,
            R.string.hide_watched,
            true
        )
        isInfinitePref = createCheckBoxPref(
            WidgetSettings.KEY_PREFIX_WIDGET_IS_INFINITE + appWidgetId,
            R.string.pref_infinite_scrolling,
            false
        )
        val isHideWatchedButtonPref = createCheckBoxPref(
            WidgetSettings.KEY_PREFIX_WIDGET_IS_HIDE_WATCH_BUTTON + appWidgetId,
            R.string.pref_hide_watch_button,
            false
        )

        // Appearance.
        themePref = createListPref(
            WidgetSettings.KEY_PREFIX_WIDGET_THEME + appWidgetId,
            R.string.pref_theme,
            R.array.widget_themes,
            R.array.widget_themes_data,
            getString(R.string.widget_theme_system)
        )
        backgroundPref = createListPref(
            WidgetSettings.KEY_PREFIX_WIDGET_BACKGROUND_OPACITY + appWidgetId,
            R.string.pref_widget_opacity,
            R.array.widgetOpacity,
            R.array.widgetOpacityData,
            WidgetSettings.DEFAULT_WIDGET_BACKGROUND_OPACITY
        )
        val isLargeFontPref = createCheckBoxPref(
            WidgetSettings.KEY_PREFIX_WIDGET_IS_LARGE_FONT + appWidgetId,
            R.string.pref_large_font,
            false
        )

        // Use settings file specific for this widget.
        preferenceManager.apply {
            sharedPreferencesName = WidgetSettings.SETTINGS_FILE
            sharedPreferencesMode = 0
        }

        // Build the preference screen.
        val prefScreen = preferenceManager.createPreferenceScreen(activity).apply {
            addPreference(typePref)
            addPreference(showsSortPref)
            addPreference(onlyFavoritesPref)
            addPreference(onlyPremieresPref)
            addPreference(onlyCollectedPref)
            addPreference(hideWatchedPreference)
            addPreference(isInfinitePref)
            addPreference(isHideWatchedButtonPref)
            val appearanceCategory = PreferenceCategory(activity).apply {
                setTitle(R.string.pref_appearance)
            }
            // Need to add to screen first so added prefs can get unique IDs.
            addPreference(appearanceCategory)
            appearanceCategory.apply {
                addPreference(themePref)
                addPreference(backgroundPref)
                addPreference(isLargeFontPref)
            }
        }
        preferenceScreen = prefScreen

        // After restoring any values prevent persisting changes (only save on user command).
        disablePersistence(prefScreen)

        // Display current value as summary.
        bindPreferenceSummaryToValue(typePref)
        bindPreferenceSummaryToValue(showsSortPref)
        bindPreferenceSummaryToValue(backgroundPref)
        bindPreferenceSummaryToValue(themePref)

        // Disable saving some prefs not available for non-supporters.
        if (!Utils.hasAccessToX(activity)) {
            val onDisablePreferenceChangeListener =
                Preference.OnPreferenceChangeListener { _: Preference?, _: Any? ->
                    Utils.advertiseSubscription(activity)
                    false
                }
            typePref.apply {
                onPreferenceChangeListener = onDisablePreferenceChangeListener
                setSummary(R.string.onlyx)
            }
            backgroundPref.apply {
                onPreferenceChangeListener = onDisablePreferenceChangeListener
                setSummary(R.string.onlyx)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        // Trigger listener to update pref states based on current type.
        preferenceChangeListener.onPreferenceChange(typePref, typePref.value)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.widget_config_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_save -> {
                saveAllPreferences() // Persistence is disabled, save manually.
                (activity as ListWidgetPreferenceActivity).updateWidget()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun createCheckBoxPref(
        key: String,
        @StringRes titleRes: Int,
        defaultValue: Boolean
    ): CheckBoxPreference {
        return CheckBoxPreference(activity).also {
            it.key = key
            it.setTitle(titleRes)
            it.setDefaultValue(defaultValue)
        }
    }

    private fun createListPref(
        key: String,
        @StringRes title: Int,
        @ArrayRes entries: Int,
        @ArrayRes values: Int,
        defaultValue: String
    ): ListPreference {
        return ListPreference(activity).also {
            it.key = key
            it.setTitle(title)
            it.setEntries(entries)
            it.setEntryValues(values)
            it.setDefaultValue(defaultValue)
            it.positiveButtonText = null
            it.negativeButtonText = null
        }
    }

    /**
     * Walks through all preferences of the [group] and sets [Preference.setPersistent] false.
     */
    private fun disablePersistence(group: PreferenceGroup) {
        val count = group.preferenceCount
        for (i in 0 until count) {
            val pref = group.getPreference(i)
            if (pref is PreferenceGroup) {
                disablePersistence(pref)
            } else {
                pref.isPersistent = false
            }
        }
    }

    private fun saveAllPreferences() {
        preferenceManager.sharedPreferences.edit {
            savePreferences(preferenceScreen, this)
        }
    }

    private fun savePreferences(
        group: PreferenceGroup,
        editor: SharedPreferences.Editor
    ) {
        val count = group.preferenceCount
        for (i in 0 until count) {
            val pref = group.getPreference(i)
            val key = pref.key
            when (pref) {
                is PreferenceGroup -> {
                    savePreferences(pref, editor)
                }
                is CheckBoxPreference -> {
                    editor.putBoolean(key, pref.isChecked)
                }
                is ListPreference -> {
                    editor.putString(key, pref.value)
                }
            }
        }
    }

    /**
     * Binds a preference's summary to its value. More specifically, when the
     * preference's value is changed, its summary (line of text below the
     * preference title) is updated to reflect the value. The summary is also
     * immediately updated upon calling this method. The exact display format is
     * dependent on the type of preference.
     */
    private fun bindPreferenceSummaryToValue(preference: ListPreference) {
        // Set the listener to watch for value changes.
        preference.onPreferenceChangeListener = preferenceChangeListener

        // Trigger the listener immediately with the preference's current value.
        preferenceChangeListener.onPreferenceChange(preference, preference.value)
    }

    private val preferenceChangeListener =
        Preference.OnPreferenceChangeListener { preference, newValue ->
            // Enable/disable settings that (do not) apply to shows, if type is shows.
            if (typePref.key == preference.key) {
                val newTypeValue = newValue as String
                val displayingShows = getString(R.string.widget_type_shows) == newTypeValue
                showsSortPref.isEnabled = displayingShows
                onlyPremieresPref.isEnabled = !displayingShows
                onlyCollectedPref.isEnabled = !displayingShows
                hideWatchedPreference.isEnabled = !displayingShows
                isInfinitePref.isEnabled = !displayingShows
            }

            // If system theme is used disable unsupported transparency setting.
            if (themePref.key == preference.key) {
                val newTypeValue = newValue as String
                val isSystemTheme = getString(R.string.widget_theme_system) == newTypeValue
                backgroundPref.isEnabled = !isSystemTheme
                if (isSystemTheme) {
                    backgroundPref.value = WidgetSettings.DEFAULT_WIDGET_BACKGROUND_OPACITY
                    // Need to manually trigger change listener for whatever reason.
                    backgroundPref.onPreferenceChangeListener.onPreferenceChange(
                        backgroundPref,
                        backgroundPref.value
                    )
                }
            }

            if (preference is ListPreference) {
                // For list preferences, look up the correct display value in
                // the preference's 'entries' list.
                val stringValue = newValue.toString()
                val index = preference.findIndexOfValue(stringValue)
                // Set the summary to reflect the new value. Escape '%'.
                preference.setSummary(
                    if (index >= 0) {
                        preference.entries[index].toString().replace("%".toRegex(), "%%")
                    } else null
                )
            }

            true
        }

    companion object {

        private const val ARG_WIDGET_ID = "appWidgetId"

        fun newInstance(appWidgetId: Int): ListWidgetPreferenceFragment {
            return ListWidgetPreferenceFragment().apply {
                arguments = bundleOf(ARG_WIDGET_ID to appWidgetId)
            }
        }

    }
}